[SYCL][ROCm] Use offload-arch instead of mcpu for AMD arch

npmiller opened the pull request:
This patch changes using `-mcpu` for SYCL applications targeting AMD to
`-Xsycl-target-backend --offload-arch`.

Before this patch the offloading arch wasn't set correctly for AMD
architectures.

This is fixing an issue with HIP that was talked about in #4133,
regarding having `v4` in the hip part of the triple, without the `v4`
HIP seems to be ignoring the fact that the offloading arch is missing
from the triple, which is why there was a workaround orignally to force
not using `v4` with SYCL. By fixing the offloading arch this patch fixes
the issue properly and now the triple with `v4` works because it also
contains the offloading architecture.

npmiller commented:
@malixian This patch should fix the issue discussed in:
* https://github.com/intel/llvm/pull/4133#issuecomment-888158380

AGindinson commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -3961,7 +3961,7 @@ class OffloadingActionBuilder final {
     ActionList FPGAArchiveInputs;
 
     /// List of CUDA architectures to use in this compilation with NVPTX targets.
```
```suggestion
    /// List of GPU architectures to use in this compilation with NVPTX/AMDGCN targets.
```

AGindinson reviewed (commented):
I believe `clang/test/Driver/sycl-offload-amdgcn.cpp` should be expanded, at least to check the error message and enforce valid command lines for the main test cases.

elizabethandrews commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -288,6 +288,8 @@ def err_drv_expecting_fsycl_with_sycl_opt : Error<
 def err_drv_fsycl_with_c_type : Error<
   "The option '%0' must not be used in conjunction with '-fsycl', which expects C++ source.">;
+def err_drv_sycl_missing_amdgpu_arch : Error<
+  "Missing AMDGPU architecture for SYCL offloading. Please specify it with -Xsycl-target-backend --offload-arch.">;
```
Clang diagnostics do not start with capital letters, and use different punctuations (See https://github.com/intel/llvm/pull/4243 for additional details). Please change the diagnostic to - 

`missing AMDGPU architecture for SYCL offloading; please specify architecture with -Xsycl-target-backend --offload-arch`

@AaronBallman did I get requirements for diagnostics correct?

AaronBallman replied:
> @AaronBallman did I get requirements for diagnostic correct?

Yes, though I have an additional suggestion here -- you should surround command line arguments with single quotes like you'd do for syntactic elements. e.g.,

`missing AMDGPU architecture for SYCL offloading; specify architecture with '-Xsycl-target-backend --offload-arch'`

(This also removed the "please" for brevity.)

npmiller replied:
I have updated the message as suggested

elizabethandrews reviewed (commented):
Please add a test for the new diagnostic

npmiller commented:
I've updated `clang/test/Driver/sycl-offload-amdgcn.cpp` to add the architecture parameters, added a test for the new diagnostic in there as well, and updated other `amdgcn` tests with architecture flags.

I'm still seeing one test failing with this when running `ninja check`, `Driver/amdgpu-openmp-toolchain.c`, however it seems that this test was already failing on AMD before this patch.

malixian commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4642,39 +4647,93 @@ class OffloadingActionBuilder final {
+        if (Triple.isAMDGCN() && llvm::none_of(GpuArchList, [&](auto &P) {
+              return P.first.isAMDGCN();
+            })) {
+          C.getDriver().Diag(clang::diag::err_drv_sycl_missing_amdgpu_arch);
```
How about adding a default AMD GPU arch?

npmiller replied:
The reason I didn't go for the default AMD GPU arch is that as I understand it we need to specify the exact GPU architecture for AMD so a default would only work for a very specific type of GPUs. Which means that in a lot of cases users would still need to specify the architecture manually, so I think it is better to force the architecture to always be set manually and have a clear diagnostic, than have a default architecture that rarely works and a more confusing error message from hip.

This is different with NVidia because `SM_50` covers a lot of different GPUs, so in most cases it will work out of the box and the user won't have to set the architecture.

AGindinson commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4642,39 +4647,93 @@ class OffloadingActionBuilder final {
+    /// Initialize the GPU architecture list from arguments - this populates
+    /// `GpuArchList` from `--offload-arch` flags. Only relevant if compiling to
+    /// CUDA or AMDGCN. Return true if any initialization errors are found.
+    /// FIXME: SPIR-V AOT targets should also use `offload-arch` to better fit
+    /// in the standard model.
```
I doubt that SPIR-V AOT targets' CLI could be conveniently switched to this usage, as the backend compiler CLIs are defined beyond the scope of the DPC++ implementation. The direction could be defined in a more "general" manner, though - along the lines of:
```suggestion
    /// FIXME: Consider providing a unified driver option for exact offload architecture
    /// so that command lines for SPIR-V AOT and CUDA/AMDGCN targets would
    /// use a common option, and the exact spelling like "offload-arch" is set within
    /// the driver.
```
(perhaps there's a more concise way to formulate this)

npmiller replied:
Yes that's true, I just meant in terms of command line interface, what about:
```suggestion
    /// FIXME: Consider defining a standard option name for "offload-arch" to
    /// use in both the SPIR-V AOT and CUDA/AMDGCN targets.
```

AGindinson replied:
Your suggestion is much better, thank you!

Naghasan replied:
This is in relation to the clang driver flags not the called tool flags (this is the purpose of the driver to do the translation).

For SPIR-V targets, the driver uses a sub arch provided with the device triple to do the processing. While this is a workable solution for the SPIR toolchain, this by-passes the existing mechanism used by clang (`BoundArch` thing) thus reinvent the wheel. This makes non SPIR toolchains harder to integrate and more prone to breakage.

npmiller replied:
Oooh I see, so what about:
```suggestion
    /// FIXME: "offload-arch" and the BoundArch mechanism should also be
    // used in the SYCLToolChain for SPIR-V AOT to get the offload
    // architecture instead of the Triple sub-arch it currently uses.
```
@Naghasan ? Does that make more sense?

Naghasan replied:
Yes, nitpicking a bit:

```suggestion
    /// FIXME: "offload-arch" and the BoundArch mechanism should also be
    // used in the SYCLToolChain for SPIR-V AOT to track the offload
    // architecture instead of the Triple sub-arch it currently uses.
```

npmiller replied:
@AGindinson are you happy with this update to the comment?

AGindinson replied:
@npmiller Works for me!

AGindinson commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -4642,39 +4647,93 @@ class OffloadingActionBuilder final {
             ParsedArg->getOption().matches(options::OPT_offload_arch_EQ)) {
+          llvm::StringRef ArchStr = ParsedArg->getValue(0);
+          if (TargetBE->isNVPTX()) {
+            // CUDA arch also applies to AMDGCN ...
+            CudaArch Arch = StringToCudaArch(ArchStr);
```
Would it make sense to rename the helper function & the architecture type as in `Cuda` -> `GPU`?

npmiller replied:
Yeah I think it would make sense, although I guess at the moment it's only AMD and NVIDIA GPUs, but `CudaArch` is definitely not quite correct since it has the AMD architectures in there too.

AGindinson replied:
I've also had this concern with "GPU" being broader than _just_ NVIDIA & AMD, but this would at least be consistent with similar names like `GpuArchList` that the patch brings in.

npmiller replied:
Is this something you'd like to see in this patch or maybe to consider in the future as a follow up?

AGindinson replied:
Personally, I'm all for a follow-up PR approach (a short GH issue would be useful just so that the issue isn't lost).

npmiller replied:
Issue filed here:
* https://github.com/intel/llvm/issues/4279

Naghasan reviewed (approved):
LGTM

elizabethandrews reviewed (approved):
LGTM

bader commented:
@AGindinson, @hchilama, @mdtoguchi, ping.

AaronBallman reviewed (approved):
LGTM

hchilama reviewed (approved):
LGTM
